api/router.js

```diff
@@ -6,31 +6,45 @@ var autoredraw = require("../api/autoredraw")
 
 module.exports = function($window, renderer, pubsub) {
 	var router = coreRouter($window)
+	var identity = function(o){return o}
+	var globalId
+	var routeRequested = false
+	function onbeforeresolve(){routeRequested = true}
+
 	var route = function(root, defaultRoute, routes) {
-		var current = {path: null, component: "div"}
+		var current = {path: null, component: "div", render:identity}
 		var replay = router.defineRoutes(routes, function(payload, args, path, route) {
-			args.path = path, args.route = route
-			if (typeof payload.onmatch === "function") {
-				if (typeof payload.view !== "function") payload.view = function(vnode) {return vnode}
-				var resolve = function(component) {
-					current.path = path, current.component = component
-					renderer.render(root, payload.view(Vnode(component, null, args, undefined, undefined, undefined)))
-				}
-				if (path !== current.path) payload.onmatch(Vnode(payload, null, args, undefined, undefined, undefined), resolve)
-				else resolve(current.component)
+			if (!routeRequested) {
+				renderer.render(root, current.render.call(payload, Vnode(current.component, null, args, undefined, undefined, undefined)))
+				return
+			}
+			routeRequested = false
+			var render = route.render || identity
+			var currentId = globalId = {}
+			var resolved = false
+			function resolve (component) {
+				if (currentId !== globalId || resolved) return
+				resolved = true
+				current.path = path, current.component = component || current.component, current.render = render
+				renderer.render(root, current.render.call(payload, Vnode(current.component, null, args, undefined, undefined, undefined)))
 			}
-			else {
-				renderer.render(root, Vnode(payload, null, args, undefined, undefined, undefined))
+			var onmatch = function(resolve){resolve(current.component)}
+			if (typeof payload.view !== "function") {
+				if (typeof payload.render === "function") render = payload.render
+				if (typeof payload.onmatch === "function") onmatch = payload.onmatch
+			} else {
+				current.component = payload
```

Comment:
@barneycarroll The five lines above mean that we either deal with a RouteResolver or a component.

If we were to add `render` and `onmatch` as class methods on components, we could instead run line 33 and 34 regardless of the presence of a view, and run `if (typeof payload.view === "function") current.component = payload`...